File: Tree/484-closestLeaf.py
from collections import defaultdict,deque
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
def closetLeaf(root,k):
    leaves=set()
    G=defaultdict(list)
    def dfs(root,par):
        if not root:
            return 
        dfs(root.left,root)
        dfs(root.right,root)
        
        if par:
            G[root.val].append(par.val)
            G[par.val].append(root.val)
        if not root.left and not root.right:
            leaves.add(root.val)
    dfs(root,None)
    q=deque([k])
    seen=set()
    while q:
        cur=q.popleft()
        if cur in leaves:
            return cur
        for n in G[cur]:
            if n not in seen:
                seen.add(n)
                q.append(n)
                
root=TreeNode(0)
one=TreeNode(1)
two=TreeNode(2)
three=TreeNode(3)
four=TreeNode(4)
root.left=one
root.right=two
one.left=three
two.left=four

print(closetLeaf(root,2))

File: Tree/484-closestLeaf.cpp
#include<map>
#include<iostream>
#include<set>
#include<deque>
#include<vector>

using VEC=std::vector<int>;
using MAP=std::map<int,VEC>;
using Q=std::deque<int>;
using SET=std::set<int>;
class TreeNode
{
    public:
        int m_val{};
        TreeNode* left{};
        TreeNode* right{};
        TreeNode(int v=0,TreeNode* l=nullptr,TreeNode* r=nullptr):m_val{v},left{l},right{r}{};
};

void foo(const TreeNode* root, const TreeNode* par,MAP& m,SET& s)
{
    if(!root)
        return ;
    foo(root->left,root,m,s);
    foo(root->right,root,m,s);
    if(par)  //double direction 
    {
        m[par->m_val].push_back(root->m_val);
        m[root->m_val].push_back(par->m_val);
    }
    if(!(root->left) &&!(root->right))
        s.insert(root->m_val);
}

int getCloestLeaf(const TreeNode* root,int k)
{
    //create the Graph for the BT
    MAP G{};
    SET leaves{};
    foo(root,nullptr,G,leaves);

    Q q{k};  //for BFS to get the shortest leaf from given target k
    // q.push_back(k);
    SET seen{};
    while(!q.empty())
    {
        int cur=q.front();
        q.pop_front();
        // auto search=leaves.find(cur);
        if(leaves.find(cur)!=leaves.end())
            return cur;
        for(auto &n:G[cur])
        {
            // auto isAlreadyAccessed=seen.find(n);
            if(seen.find(n)==seen.end())
            {
                seen.insert(n);
                q.push_back(n);
            }
        }
    }
}


int main()
{
    TreeNode* root=new TreeNode();
    TreeNode* one=new TreeNode(1);
    TreeNode* two=new TreeNode(2);
    TreeNode* three=new TreeNode(3);
    TreeNode* four=new TreeNode(4);

    root->left=one;
    root->right=two;
    one->left=three;
    two->right=four;

    int leaf=getCloestLeaf(root,1);
    std::cout<<leaf;

}